### PR TITLE
Implemented the PE3 ECU library

### PR DESCRIPTION
**Pull request by crsz20**
### Summary
The final piece to integrating our ECU and CAN interfaces/libraries.

The `Pe3` library does **not** care about how the hardware is configured for reading data. It is up to the user to manage configuring the hardware for enabling polling or interrupt modes. Hardware configuration is done through the `bx_can_peripheral` pointer. However, `Pe3` interacts with it through the abstract interface `ICan` to be dependent on a CAN bus while being platform-independent.

There are a total of 3 pointers that are pointing to stm32's BxCAN hardware peripheral. For this reason, the peripheral is managed using `shared_ptr` instead of `unique_ptr`.

### Checks
- [x] Tested

### CODE REVIEW

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/iecu.hpp [hunk 1]
```diff
@@ -12,11 +12,23 @@
...
+
+	virtual const std::vector<uint32_t>& CanIdList() = 0;
+
+	virtual uint32_t LatestCanId() = 0;
+
+	virtual bool NewMessageArrived() = 0;
+
+	virtual void Update() = 0;
+
+	// TODO: getter methods for each variable
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/ECU/PE3/iecu.hpp]: Will remove this in my next commit.

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp [hunk 2]
```diff
@@ -14,12 +14,186 @@
...
+
+	// Frame PE8 Fields
+	float rpm_rate_ = 0.0f;
+	float tps_rate_ = 0.0f;
+	float map_rate_ = 0.0f;
+	float maf_load_rate_ = 0.0f;
+
+	// Frame PE9 Fields
+	float measured_lambda_1_ = 0.0f;
+	float measured_lambda_2_ = 0.0f;
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp]: I find it a little difficult justifying an array of just two elements (I would also initialize a constant for its size), but I have also considered using C++ [std::pair](https://www.geeksforgeeks.org/pair-in-cpp-stl/) since this is enumerated `1` and `2` on the datasheet.

Probably overkill — but I have no preference. I will leave it up to you.

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp [hunk 3]
```diff
@@ -14,12 +14,186 @@
...
+
+	// Frame PE14 Fields
+	float fuel_comp_acceleration_ = 0.0f;
+	float fuel_comp_starting_ = 0.0f;
+	float fuel_comp_air_temp_ = 0.0f;
+	float fuel_comp_coolant_temp_ = 0.0f;
+
+	// Frame PE15 Fields
+	float fuel_comp_barometer_ = 0.0f;
+	float fuel_comp_map_ = 0.0f;
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp:267]: I am considering creating a struct. Something along the lines of...

```c++
struct FuelCompensation {
  acceleration
  starting
  air_temp
  coolant temp
  barometer
  comp map
}

```

If so, I may do a similar approach for `ignition_comp_...`.

Thoughts?

Comment (WilliamLim781) [reply Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp:267]: yeah that doesn't sound like a bad idea for these types of frame fields

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp (hunk 2 above)
Comment (WilliamLim781) [reply Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp]: I'm honestly fine with whatever. Pick whatever makes the most convenience for you

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/iecu.hpp (hunk 1 above)
Comment (WilliamLim781) [reply Project/DAQ_System/Program/Src/Sensor/ECU/PE3/iecu.hpp]: 👍

**Review by WilliamLim781 — APPROVED** [review]
🧇

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp [hunk 4]
```diff
@@ -14,12 +14,186 @@
...
+#define FramePe9Id		0x0CFFF848
+#define FramePe10Id		0x0CFFF948
+#define FramePe11Id		0x0CFFFA48
+#define FramePe12Id		0x0CFFFB48
+#define FramePe13Id		0x0CFFFC48
+#define FramePe14Id		0x0CFFFD48
+#define FramePe15Id		0x0CFFFE48
+#define FramePe16Id		0x0CFFD048
+
+class Pe3 : IEcu {
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp]: Caught a bug. I left out the `public` keyword when inheriting the `IEcu` interface. This was giving me issue with pointer arithmetic at initialization. Here is an [example](https://www.reddit.com/r/cpp_questions/comments/18un06p/error_error_conversion_from_shared_ptr_to/).

**Review by alexshidagoatnocap — APPROVED** [review]

File: Project/DAQ_System/Program/Src/app.cpp [hunk 5]
```diff
@@ -133,30 +111,32 @@ void cppMain() {
...
 
-			} else if(can_bus->LatestCanId() == 0x0CFFF548){
-				printf("\t\t 0x0CFFF548 \n");
-				printf("\r");
-				// Process raw data
+
+			case FramePe6Id:
+				printf("\t\t %" PRIu32 "\n", can_id);
+				break;
 			}
```
Comment (crsz20) [inline Project/DAQ_System/Program/Src/app.cpp]: As I develop integrate the `IEcu` interface, I am starting to realize that in practice we may interact with the `Pe3` ECU class directly (as shown in this snippet).

Interacting through `IEcu`, we will lose the ability to identify which data from ECU got updated:
```c++

if (generic_ecu->NewMessageArrived()) {
  generic_ecu->Update();
  // get data
}
```

Notice that in this example snippet, I am not using the switch statement to handle specific CAN IDs. The IDs are ECU-specific. So, although I can use `generic_ecu->LatestCanId()` to retrieve the ID, identifying it will require ECU-specific knowledge that is held in `Pe3`.

The example snippet still works, but identifying the CAN message will be more efficient. The example is essentially polling `Pe3` through `IEcu`. On the other hand, with the switch statement of ECU-specific CAN IDs, we would only be retrieving the data that was updated — instead of retrieving all of it.

**Long story short: `IEcu` would only support polling. While using `Pe3` directly will allow us to reap the benefits of an interrupt approach.**

So, in practice, I think we will have the DAQ use `Pe3` directly, and maintain `IEcu` for aiding developers to implement the minimum requirements of an ECU. But, the DAQ firmware will not interact to the ECU component using `IEcu`, only its concrete classes.

Comment (crsz20) [reply Project/DAQ_System/Program/Src/app.cpp]: A solution may be to implement the [Observer design pattern](https://www.modernescpp.com/index.php/the-observer-pattern/). But I think that outside the scope of this task and there are higher priorities.

Comment (crsz20) [reply Project/DAQ_System/Program/Src/app.cpp]: Updated `Pe3` usage, using its ecu-specific CAN IDs to act upon only the data that was recently updated:

image

Updated `IEcu` usage, where CAN ID identification is lost — so everything must be polled:

image

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp (hunk 2 above)
Comment (crsz20) [reply Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp]: I just used `std::array` (not be confused with raw arrays) for their automatic bounds-checking at compile-time.

File: Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp (hunk 3 above)
Comment (crsz20) [reply Project/DAQ_System/Program/Src/Sensor/ECU/PE3/pe3.hpp:267]: Conclusion: decided not to use a struct because it will be syntactically and visually similar.

**Review by WilliamLim781 — APPROVED** [review]
